frame/contracts/src/rent.rs

```diff
@@ -14,102 +14,152 @@
+	/// Remove the contract along with it's storage.
+	Kill,
+	/// The contract cannot afford payment within its rent budget so it gets evicted. However,
+	/// because its balance is greater than the subsistence threshold it leaves a tombstone.
+	Evict {
+		amount: Option<OutstandingAmount<T>>,
+	},
+	/// Everything is OK, we just only take some charge.
+	Charge { amount: OutstandingAmount<T> },
 }
 
-/// Evict and optionally pay dues (or check account can pay them otherwise) at the current
-/// block number (modulo `handicap`, read on).
+/// Returns a fee charged per block from the contract.
 ///
-/// `pay_rent` gives an ability to pay or skip paying rent.
-/// `handicap` gives a way to check or pay the rent up to a moment in the past instead
-/// of current block.
+/// This function accounts for the storage rent deposit. I.e. if the contract possesses enough funds
+/// then the fee can drop to zero.
+fn compute_fee_per_block<T: Trait>(
+	balance: &BalanceOf<T>,
+	contract: &AliveContractInfo<T>,
+) -> BalanceOf<T> {
+	let free_storage = balance
+		.checked_div(&T::RentDepositOffset::get())
+		.unwrap_or_else(Zero::zero);
+
+	let effective_storage_size =
+		<BalanceOf<T>>::from(contract.storage_size).saturating_sub(free_storage);
+
+	effective_storage_size
+		.checked_mul(&T::RentByteFee::get())
+		.unwrap_or(<BalanceOf<T>>::max_value())
+}
+
+/// Subsistence threshold is the extension of the minimum balance (aka existential deposit) by the
+/// tombstone deposit, required for leaving a tombstone.
 ///
-/// NOTE: This function acts eagerly, all modification are committed into the storage.
-fn try_evict_or_and_pay_rent<T: Trait>(
-	account: &T::AccountId,
-	handicap: T::BlockNumber,
-	pay_rent: bool,
-) -> (RentOutcome, Option<ContractInfo<T>>) {
-	let contract_info = <ContractInfoOf<T>>::get(account);
-	let contract = match contract_info {
-		None | Some(ContractInfo::Tombstone(_)) => return (RentOutcome::Exempted, contract_info),
-		Some(ContractInfo::Alive(contract)) => contract,
-	};
+/// Rent mechanism cannot make the balance lower than subsistence threshold.
+fn subsistence_threshold<T: Trait>() -> BalanceOf<T> {
+	T::Currency::minimum_balance() + T::TombstoneDeposit::get()
+}
 
-	let current_block_number = <frame_system::Module<T>>::block_number();
+/// Returns amount of funds available to consume by rent mechanism.
+///
+/// Rent mechanism cannot consume more than `rent_allowance` set by the contract and it cannot make
+/// the balance lower than [`subsistence_threshold`].
+fn rent_budget<T: Trait>(
+	balance: &BalanceOf<T>,
+	contract: &AliveContractInfo<T>,
+	subsistence_threshold: &BalanceOf<T>,
+) -> BalanceOf<T> {
+	let rent_allowed_to_charge = *balance - *subsistence_threshold;
+	<BalanceOf<T>>::min(contract.rent_allowance, rent_allowed_to_charge)
+}
 
+/// Consider the case for rent payment of the given account and returns a `Verdict`.
+///
+/// The `current_block_number` must be equal to the current block number. Use `handicap` do
```

Comment:
> The `current_block_number` must be equal to the current block number.

Sounds pretty obvious :D

Reply:
Ahh, good point. I will rephrase.

This is also from https://github.com/paritytech/substrate/pull/4733 . I will file a separate PR to address this.

Reply:
https://github.com/paritytech/substrate/pull/4790